const mongoose = require('mongoose');
const dotenv = require('dotenv');

process.on('uncaughtException', err => { // catch all the errors from synchronous function not from promise
  console.log('UNCAUGHT EXCEPTION! ðŸ’¥ Shutting down...');
  console.log(err.name, err.message);
  process.exit(1);
});

dotenv.config({ path: './config.env' });
const app = require('./app');

const DB = process.env.DATABASE.replace(
  '<PASSWORD>',
  process.env.DATABASE_PASSWORD
);

mongoose
  .connect(DB, {
    useNewUrlParser: true,
    useCreateIndex: true,
    useFindAndModify: false
  })
  .then(() => console.log('DB connection successful!'));

const port = process.env.PORT || 3000;
const server = app.listen(port, () => {
  console.log(`App running on port ${port}...`);
});

process.on('unhandledRejection', err => { // all ohter promise rejection error will be handelled
  console.log('UNHANDLED REJECTION! ðŸ’¥ Shutting down...');
  console.log(err.name, err.message);
  server.close(() => {
    process.exit(1);
  });
});

process.on('SIGTERM', () => {                                      // SIGTERM is a signal that is used to cause a program stop running in a polite way
  console.log('ðŸ‘‹ SIGTERM RECEIVED. Shutting down gracefully');
  server.close(() => {
    console.log('ðŸ’¥ Process terminated!');
  });
});
